remove unnecessary sync point in AveragedModel update

gl3lan opened the pull request:
Summary:
The test `bool(self.n_averaged == 0)` is a CPU/GPU synchronization point that is called for each update.
This test is only meant to know whether the AveragedModel copy has been initialized or not.
This diff introduces a CPU-based variable for that purpose.
When loading from checkpoint we also make sure the parameter is refreshed.

After this fix, each `update_parameter` call is reduced to 6ms from 333ms (98% reduction).

Test Plan:
contbuild & OSS CI
Test plan from GitHub:
CI

Rollback Plan:

Differential Revision: D78074709

janeyx99 reviewed (commented):
Huh why is n_averaged even a tensor. It can just be a python number and this sync would go away, right?

gl3lan commented:
> I've just learned from @mikaylagawarecki that there is a get_extra_state and set_extra_state on modules that we can use to just store a Python number for n_averaged: https://docs.pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.get_extra_state
> 
> This would make this code simpler and more understandable with the perf win if we had n_averaged as a python number instead of buffer

@janeyx99 it looks cleaner now but it breaks any pre-existing checkpoint loading because of the missing buffer. Any suggestion?

janeyx99 commented:
@gl3lan yea we should be able to register a hook https://docs.pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_load_state_dict_pre_hook so that if the state dict has a Tensor n_averaged, we just convert it to the number. How does that sound to you?

I'm also okay with the overriding load_state_dict option, whichever is simpler. Can you also add a test case to ensure we don't break existing users?

gl3lan committed:
remove unnecessary sync point in AveragedModel update #suppress-bc-linter (#158017)

Summary:

The test `bool(self.n_averaged == 0)` is a CPU/GPU synchronization point that is called for each update.
This test is only meant to know whether the AveragedModel copy has been initialized or not.
This diff introduces a CPU-based boolean variable for that purpose.
When loading from checkpoint we also make sure the parameter is refreshed.

After this fix, each `update_parameter` call is reduced to 6ms from 333ms (98% reduction).

Test Plan:
contbuild & OSS CI
Test plan from GitHub:
CI

Rollback Plan:

Differential Revision: D78074709

janeyx99 reviewed (dismissed):
thank u!

janeyx99 commented:
@pytorchbot merge
